File: web/src/lib/api/permits.ts
import { supabase } from "@/lib/supabase";

// P3 Leistenschneider — typed client for the work-permit validation API.
// Own module by design; do NOT fold into the shared lib/api.ts.

const API = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

async function authHeaders(): Promise<Record<string, string>> {
  if (!supabase) return {}; // dev mode (AUTH_MODE=dev) — no token needed
  const { data } = await supabase.auth.getSession();
  const token = data.session?.access_token;
  return token ? { Authorization: `Bearer ${token}` } : {};
}

export type Decision = "VALID" | "EXPIRED" | "NOT_A_PERMIT" | "NOT_WORK_AUTHORIZED" | "NEEDS_REVIEW";

export interface RubricItem {
  label: string;
  weight: number;
  earned: number;
  grounded: boolean | null;
  detail: string;
}

export interface PermitFields {
  is_work_permit: boolean;
  document_type: string | null;
  document_type_quote: string | null;
  holder_name: string | null;
  nationality: string | null;
  legal_basis: string | null;
  issue_date: string | null;
  valid_until: string | null;
  valid_until_quote: string | null;
  employment_allowed: boolean | null;
  employment_quote: string | null;
  extraction_confidence: number;
  notes: string | null;
}

export interface PermitCheck {
  id: string;
  filename: string | null;
  decision: Decision;
  confidence: number;
  valid_until: string | null;
  days_remaining: number | null;
  employment_status: string | null;
  holder_name: string | null;
  document_type: string | null;
  legal_basis: string | null;
  legal_basis_citation: string | null;
  needs_review: boolean;
  status: string; // pending | confirmed | overridden
  fields: PermitFields;
  rubric: RubricItem[];
  reasons: string[];
  created_at: string;
}

export interface PermitHistoryRow {
  id: string;
  filename: string | null;
  decision: Decision;
  confidence: number;
  valid_until: string | null;
  employment_status: string | null;
  holder_name: string | null;
  needs_review: boolean;
  status: string;
  created_at: string;
}

export interface ReviewActionRow {
  id: string;
  reviewer: string;
  outcome: string;
  override_decision: string | null;
  note: string | null;
  created_at: string;
}

async function req<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API}${path}`, {
    ...init,
    headers: { ...(init?.headers ?? {}), ...(await authHeaders()) },
  });
  if (!res.ok) throw new Error(`API ${res.status}: ${await res.text()}`);
  return res.json() as Promise<T>;
}

export async function postPermit(file: File): Promise<PermitCheck> {
  const fd = new FormData();
  fd.append("file", file);
  return req<PermitCheck>("/agents/permits", { method: "POST", body: fd });
}

export async function getPermitHistory(): Promise<PermitHistoryRow[]> {
  return req<PermitHistoryRow[]>("/agents/permits/history");
}

export async function getReviewQueue(): Promise<PermitCheck[]> {
  return req<PermitCheck[]>("/agents/permits/review-queue");
}

export async function getPermit(id: string): Promise<PermitCheck> {
  return req<PermitCheck>(`/agents/permits/${id}`);
}

export async function reviewPermit(
  id: string,
  body: { outcome: "confirmed" | "overridden"; override_decision?: Decision; note?: string },
): Promise<PermitCheck> {
  return req<PermitCheck>(`/agents/permits/${id}/review`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

File: services/permits.py
"""P3 Leistenschneider — Work-Permit Validation service (tenant-scoped, persisted, audited).

Pipeline:
  1. INGEST   — uploaded bytes -> content blocks (PDF/image read NATIVELY, no OCR).
  2. VALIDATE — agents.permits.validate_blocks: one structured pass -> grounded fields ->
                deterministic verdict + itemized rubric + §AufenthG consult.
  3. PERSIST  — PermitCheck (+ full grounded payload) + AuditLog, scoped to the tenant.
  4. OVERSIGHT— below-threshold / implied-by-statute checks (needs_review) sit in a
                human-review queue; a reviewer confirm/override is recorded in ReviewAction.
                The system never auto-issues a binding decision; the human can always override.

agents/permits.py stays PURE logic; all persistence lives here.
"""
from __future__ import annotations

from datetime import date
from pathlib import Path

from sqlmodel import Session, desc, select

from agents import permits
from agents.permits import PermitResult
from core import ingest
from core.db import engine
from core.models import AuditLog, PermitCheck, ReviewAction

_DECISIONS = {"VALID", "EXPIRED", "NOT_A_PERMIT", "NOT_WORK_AUTHORIZED", "NEEDS_REVIEW"}


# --------------------------------------------------------------------------- #
# process: ingest -> validate -> persist
# --------------------------------------------------------------------------- #
def process(data: bytes, filename: str, *, tenant_id: str,
            today: date | None = None, provider: str | None = None) -> dict:
    blocks = ingest.bytes_to_blocks(data, Path(filename).suffix, filename)
    result = permits.validate_blocks(blocks, today=today, provider=provider)
    return save_check(tenant_id, filename, result)


def save_check(tenant_id: str, filename: str | None, r: PermitResult) -> dict:
    """Persist a validated result + audit it. Returns the stored check as a dict."""
    with Session(engine) as s:
        check = PermitCheck(
            tenant_id=tenant_id, filename=filename,
            decision=r.decision, confidence=r.confidence, valid_until=r.valid_until,
            days_remaining=r.days_remaining, employment_status=r.employment_status,
            holder_name=r.fields.holder_name, document_type=r.fields.document_type,
            legal_basis=r.fields.legal_basis, legal_basis_citation=r.legal_basis_citation,
            needs_review=r.needs_review, status="pending",
            fields=r.fields.model_dump(), rubric=[i.model_dump() for i in r.rubric], reasons=r.reasons,
        )
        s.add(check)
        s.add(AuditLog(tenant_id=tenant_id, action="permit.checked", severity="info",
                       detail={"decision": r.decision, "confidence": r.confidence,
                               "valid_until": r.valid_until, "needs_review": r.needs_review}))
        if r.needs_review:
            s.add(AuditLog(tenant_id=tenant_id, action="permit.review_required", severity="warning",
                           detail={"confidence": r.confidence, "reasons": r.reasons}))
        s.commit()
        s.refresh(check)
        return _check_dict(check)


# --------------------------------------------------------------------------- #
# read: history, review queue, single check
# --------------------------------------------------------------------------- #
def history(tenant_id: str, limit: int = 20) -> list[dict]:
    with Session(engine) as s:
        rows = s.exec(select(PermitCheck).where(PermitCheck.tenant_id == tenant_id)
                      .order_by(desc(PermitCheck.created_at)).limit(limit)).all()
        return [_check_summary(c) for c in rows]


def review_queue(tenant_id: str, limit: int = 50) -> list[dict]:
    with Session(engine) as s:
        rows = s.exec(select(PermitCheck)
                      .where(PermitCheck.tenant_id == tenant_id, PermitCheck.needs_review == True,  # noqa: E712
                             PermitCheck.status == "pending")
                      .order_by(desc(PermitCheck.created_at)).limit(limit)).all()
        return [_check_dict(c) for c in rows]


def get_check(tenant_id: str, check_id: str) -> dict | None:
    with Session(engine) as s:
        c = s.get(PermitCheck, check_id)
        return _check_dict(c) if c and c.tenant_id == tenant_id else None


def review_history(tenant_id: str, check_id: str) -> list[dict]:
    with Session(engine) as s:
        rows = s.exec(select(ReviewAction)
                      .where(ReviewAction.tenant_id == tenant_id, ReviewAction.permit_check_id == check_id)
                      .order_by(desc(ReviewAction.created_at))).all()
        return [{"id": a.id, "reviewer": a.reviewer, "outcome": a.outcome,
                 "override_decision": a.override_decision, "note": a.note,
                 "created_at": a.created_at.isoformat()} for a in rows]


# --------------------------------------------------------------------------- #
# oversight: confirm / override (human-in-the-loop)
# --------------------------------------------------------------------------- #
def review(tenant_id: str, check_id: str, *, reviewer: str, outcome: str,
           override_decision: str | None = None, note: str | None = None) -> dict | None:
    """Record a human decision. outcome='confirmed' keeps the recommendation;
    'overridden' replaces the decision with override_decision. Always audited."""
    if outcome not in ("confirmed", "overridden"):
        raise ValueError("outcome must be 'confirmed' or 'overridden'")
    if outcome == "overridden" and override_decision not in _DECISIONS:
        raise ValueError(f"override_decision must be one of {sorted(_DECISIONS)}")
    with Session(engine) as s:
        check = s.get(PermitCheck, check_id)
        if not check or check.tenant_id != tenant_id:
            return None
        if outcome == "overridden":
            check.decision = override_decision
            check.status = "overridden"
        else:
            check.status = "confirmed"
        check.needs_review = False
        s.add(check)
        s.add(ReviewAction(tenant_id=tenant_id, permit_check_id=check_id, reviewer=reviewer,
                           outcome=outcome, override_decision=override_decision, note=note))
        s.add(AuditLog(tenant_id=tenant_id, action="permit.reviewed", severity="info",
                       detail={"check_id": check_id, "reviewer": reviewer, "outcome": outcome,
                               "override_decision": override_decision}))
        s.commit()
        s.refresh(check)
        return _check_dict(check)


# --------------------------------------------------------------------------- #
# serialization
# --------------------------------------------------------------------------- #
def _check_summary(c: PermitCheck) -> dict:
    return {"id": c.id, "filename": c.filename, "decision": c.decision, "confidence": c.confidence,
            "valid_until": c.valid_until, "employment_status": c.employment_status,
            "holder_name": c.holder_name, "needs_review": c.needs_review, "status": c.status,
            "created_at": c.created_at.isoformat()}


def _check_dict(c: PermitCheck) -> dict:
    return {**_check_summary(c), "days_remaining": c.days_remaining,
            "document_type": c.document_type, "legal_basis": c.legal_basis,
            "legal_basis_citation": c.legal_basis_citation, "fields": c.fields,
            "rubric": c.rubric, "reasons": c.reasons}
